src/Adapter/MSTest.TestAdapter/Execution/UnitTestRunner.cs

```diff
@@ -288,33 +289,40 @@ private class ClassCleanupManager
         private readonly ClassCleanupBehavior? _lifecycleFromMsTest;
         private readonly ClassCleanupBehavior _lifecycleFromAssembly;
         private readonly ReflectHelper _reflectHelper;
-        private readonly Dictionary<string, HashSet<string>> _remainingTestsByClass;
+        private readonly ConcurrentDictionary<string, HashSet<string>> _remainingTestsByClass;
 
         public ClassCleanupManager(
             IEnumerable<UnitTestElement> testsToRun,
             ClassCleanupBehavior? lifecycleFromMsTest,
             ClassCleanupBehavior lifecycleFromAssembly,
             ReflectHelper? reflectHelper = null)
         {
-            _remainingTestsByClass = testsToRun.GroupBy(t => t.TestMethod.FullClassName)
-                .ToDictionary(
-                    g => g.Key,
-                    g => new HashSet<string>(g.Select(t => t.TestMethod.UniqueName)));
+            _remainingTestsByClass =
+                new(testsToRun.GroupBy(t => t.TestMethod.FullClassName)
+                    .ToDictionary(
+                        g => g.Key,
+                        g => new HashSet<string>(g.Select(t => t.TestMethod.UniqueName))));
             _lifecycleFromMsTest = lifecycleFromMsTest;
             _lifecycleFromAssembly = lifecycleFromAssembly;
             _reflectHelper = reflectHelper ?? new ReflectHelper();
         }
 
-        public void MarkTestComplete(TestMethodInfo testMethodInfo, TestMethod testMethod, out bool shouldRunEndOfClassCleanup, out bool shouldRunEndOfAssemblyCleanup)
+        public void MarkTestComplete(TestMethodInfo testMethodInfo, TestMethod testMethod, out bool shouldRunEndOfClassCleanup,
+            out bool shouldRunEndOfAssemblyCleanup)
         {
             shouldRunEndOfClassCleanup = false;
-            var testsByClass = _remainingTestsByClass[testMethodInfo.TestClassName];
+            shouldRunEndOfAssemblyCleanup = false;
+            if (!_remainingTestsByClass.TryGetValue(testMethodInfo.TestClassName, out var testsByClass))
```

Comment:
Is it possible 2 "complete" for same unique test?

Reply:
This is now using the method full signature (namespace + class + method sig + args) so I don't expect any duplicate.